Extend `PortfolioOptimization` to support integer variables

a-matsuo opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Resolves #54. 

### Details and comments
Currently, `PortfolioOptimization` supports only binary variables which means we can only chose whether we buy an asset or not. The PR extends `PortfolioOptimization` to support integer variables. So, now, we can decide how many asset we buy instead of just tow choices, buy or not.

stefan-woerner reviewed (approved):
lgtm

woodsp-ibm commented on qiskit_finance/applications/optimization/portfolio_optimization.py:
```diff
@@ -174,3 +217,22 @@ def budget(self, budget: int) -> None:
+        Args:
+            bounds: The lower bounds and upper bounds of each assets selectable
+        """
+        self._check_compatibility()  # check compatibility before setting bounds
+        self._bounds = bounds
```
Am I missing something. The compatibility check seems to check want is in self._bounds, which is the current bounds prior to setting in this new value. I would understand if bounds was passed to check_compatibility so whatever was passed was checked before the instance var was set, but thats not done.

a-matsuo replied:
Thank you! Apparently, my eyes were not seeing...xD

woodsp-ibm commented:
Since this is a new feature for PortfolioOptimization can we add a reno for this please.

woodsp-ibm commented on qiskit_finance/applications/optimization/portfolio_optimization.py:
```diff
@@ -103,6 +117,35 @@ def interpret(self, result: Union[OptimizationResult, np.ndarray]) -> List[int]:
+                )
+            if any(ub_ < lb_ for lb_, ub_ in self._bounds):
+                raise QiskitFinanceError(
+                    "The upper bound of each variable must be larger than the lower bound of that.",
```
```suggestion
                    "The upper bound of each variable in the list must be larger than its lower bound.",
```
Just a suggestion that may make it a little easier to read. 

I have a question too. Do we not support the case where the upper and lower bound are the same - maybe that fixes the value so it should be taken out of the optimization problem.

a-matsuo replied:
Thank you! I fixed the message. 

The current implementation supports the case where the upper and lower bound are the same. It raises an error only when the upper bound of some variable is lower than its lower bound.

woodsp-ibm replied:
Yes in looking at the code that is true - the error message states though that it "must be larger than.." which would not be true if they were equal and that is supported. Wwould it therefore more correct to say "must be greater than or equal to the lower bound" in the error message.

a-matsuo replied:
Ah, you're right. the error message is inappropriate. I fixed it. Thx!

a-matsuo commented:
@woodsp-ibm Thank you always for helpful comments. I fixed `_check_compatibility` and the error message. I also added reno (I often forget about it). Could you please review the PR again?

woodsp-ibm commented on releasenotes/notes/portfolio_intger-f88d116e5b40c6ad.yaml:
```diff
@@ -0,0 +1,6 @@
+features:
+  - |
+    :class`qiskit_finance.applications.optimization.PortfolioOptimization` now supports
+    integer variables. So, users can decide how many asset they buy instead of just tow choices,
```
```suggestion
    integer variables. So, users can decide how many asset they buy instead of just two choices,
```

woodsp-ibm commented on releasenotes/notes/portfolio_intger-f88d116e5b40c6ad.yaml:
```diff
@@ -0,0 +1,6 @@
+features:
+  - |
+    :class`qiskit_finance.applications.optimization.PortfolioOptimization` now supports
+    integer variables. So, users can decide how many asset they buy instead of just two choices,
```
```suggestion
    integer variables. So, users can decide how many assets they buy instead of just two choices,
```

woodsp-ibm reviewed (approved):
I just fixed a minor thing in reno to make 'assets' plural as it reads easier like that.

All looks good now. Thanks
